torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -105,6 +110,116 @@ def foreach_all_gather_copy_out(
         torch._foreach_copy_(out, splits)  # one `copy_` per parameter
 
 
+@torch.no_grad()
+def foreach_reduce_scatter(
+    fsdp_params: List[FSDPParam],
+    unsharded_grads: List[torch.Tensor],
+    group: dist.ProcessGroup,
+    reduce_scatter_stream: torch.cuda.Stream,
+    orig_dtype: torch.dtype,
+    reduce_dtype: Optional[torch.dtype],
+    device: torch.device,
+    predivide_factor: float,
+    postdivide_factor: float,
+) -> torch.cuda.Event:
+    """
+    ``unsharded_grads`` owns the references to the gradients computed by
+    autograd, so clearing the list frees the gradients.
+    """
+    grad_dtypes = {grad.dtype for grad in unsharded_grads}
+    if len(grad_dtypes) != 1:
+        # Check this at runtime since it could be a real runtime error if e.g.
+        # fp8 weights do not produce the correct higher precision gradients
+        _raise_assert_with_print(
+            f"FSDP reduce-scatter expects uniform gradient dtype but got {grad_dtypes}"
+        )
+    grad_dtype = unsharded_grads[0].dtype
+    reduce_dtype = reduce_dtype or grad_dtype
+    world_size = group.size()
+    padded_unsharded_sizes = tuple(
+        _get_dim0_padded_size(grad.size(), world_size) for grad in unsharded_grads
+    )
+    reduce_scatter_input_numel = sum(s.numel() for s in padded_unsharded_sizes)
+    reduce_scatter_output_numel = reduce_scatter_input_numel // world_size
+    current_stream = torch.cuda.current_stream()
+    reduce_scatter_stream.wait_stream(current_stream)
+    with torch.cuda.stream(reduce_scatter_stream):
+        reduce_scatter_input = torch.empty(
+            (reduce_scatter_input_numel,), dtype=reduce_dtype, device=device
+        )
+        foreach_reduce_scatter_copy_in(
+            fsdp_params, unsharded_grads, reduce_scatter_input, world_size
+        )
+        _div_if_needed(reduce_scatter_input, predivide_factor)
+        # Record to mark the end of the reduce-scatter copy-in in the RS stream
+        copy_in_event = torch.cuda.Event()
+        copy_in_event.record()
+        reduce_scatter_output = torch.empty(
```

Comment:
nit: I changed this to `reduce_scatter_input.new_empty((reduce_scatter_output_numel,))`.